design-proposals/klip-59-json-functions.md

```diff
@@ -0,0 +1,244 @@
+# KLIP-59 - JSON functions
+
+**Authors**: Aleksandr Sorokoumov (@gerrrr), Colin Hicks (@colinhicks)
+**Release Target**: TBD
+**Status**: _In Discussion_ |
+**Discussion**: [GitHub PR](https://github.com/confluentinc/ksql/pull/8550)
+
+## Motivation and background
+
+Many ksqlDB users leverage JSON. ETL pipelines often deal in large JSON blobs from, e.g., MongoDB.
+As the de-facto format of the web, it is often the format of analytics events produced to Kafka, as
+well as a choice for materialized views consumed by web apps.
+
+We want to position ksqlDB as a JSON-friendly database. Support for reading, writing, and
+traversing data as JSON presents an important value-add. And users of traditional databases like
+PostgreSQL already enjoy powerful support for JSON.
+
+Today, ksqlDB can:
+* Read JSON-formatted keys and values in input records.
+* Serialize keys and values as JSON in output records.
+* Extract a string value at a given JSON path.
+
+
+## What is in scope
+
+* Validate if a given string contains JSON value
+* Calculate the length of a JSON array encoded in a string
+* Extract keys and records from a given JSON-formatted string
+* Concatenate JSON-formatted strings
+* Convert a ksqlDB record to JSON-formatted string
+
+## What is not in scope
+
+* A JSON datatype
+* Binary JSON support
+
+## Public APIs
+
+Users can use ksqlDB’s varchar type on columns that they know are JSON formatted. A JSON-formatted
+string can also be nested within a map or struct.
+
+### is_json
+
+```
+is_json(json_string)
+```
+
+Given a string, returns true if it can be parsed as a valid JSON value, false otherwise.
+
+#### Examples
+
+```
+is_json("[1, 2, 3]") // returns true
+is_json("{}") // returns true
+is_json("1") // returns true
+is_json("\"abc\"") // returns true
+is_json("null") // returns true
```

Comment:
For my own education. What is `null` in JSON? Is the an "empty" JSON (is it considered a primitive or complex type)?

Reply:
JSON `null` is a primitive type. See [RFC-4627](https://datatracker.ietf.org/doc/html/rfc4627):

> JSON can represent four primitive types (strings, numbers, booleans,
>   and null) and two structured types (objects and arrays).